remote-trait-object/src/packet.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::fmt;
+
+pub struct SlotId(u32);
+
+impl fmt::Debug for SlotId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SlotId").field("id", &self.0).field("type", &self.get_type()).finish()
+    }
+}
+
+#[derive(Debug)]
+pub enum SlotType {
+    Request,
+    Response,
+}
+
+impl SlotId {
+    pub fn empty() -> Self {
+        Self(999)
+    }
+
+    pub fn new(num: u32) -> Self {
+        Self(num)
+    }
+
+    // FIXME: Please use unique slot id
+    pub fn new_request() -> Self {
+        Self(1001)
+    }
+
+    pub fn get_type(&self) -> SlotType {
+        if self.0 >= SLOT_CALL_OR_RETURN_INDICATOR.0 {
+            SlotType::Request
+        } else {
+            SlotType::Response
+        }
+    }
+
+    pub fn change_to_response(&mut self) {
+        assert!(self.0 >= SLOT_CALL_OR_RETURN_INDICATOR.0);
+        self.0 -= SLOT_CALL_OR_RETURN_INDICATOR.0;
+    }
+}
+
+const SLOT_CALL_OR_RETURN_INDICATOR: SlotId = SlotId(1000);
+
+// FIXME: repr(C) is not a reliable encoding method.
+// We need to fix the endianness of binary data.
+#[repr(C)]
+struct PacketHeader {
+    pub slot: SlotId,
+    pub target_service_name: [u8; 100],
+    pub method: [u8; 100],
```

Comment:
Please put `FIXME: use an integer index for the service and method`.

Reply:
I added a comment.